File: IwashiScopePackage/Sources/IwashiScopeFeature/Models/ColorAppearance/CIECAM16.swift
import Foundation

enum ColorAppearanceError: Error, Equatable, LocalizedError, Sendable {
    case invalidViewingConditions
    case invalidStimulus
    case outsideModelDomain
    case insufficientSpectrum

    var errorDescription: String? {
        switch self {
        case .invalidViewingConditions: String(localized: "色の見え方の観察条件を計算できません。")
        case .invalidStimulus: String(localized: "色の見え方を計算する測色値が不正です。")
        case .outsideModelDomain: String(localized: "この色は色の見え方モデルの計算範囲外です。")
        case .insufficientSpectrum: String(localized: "比較に必要な波長範囲のスペクトルが不足しています。")
        }
    }
}

/// White XYZ and Yb use the Yw = 100 scale. LA is an absolute luminance, not XYZ Y.
struct AppearanceViewingConditions: Equatable, Sendable {
    let whiteXYZ: Vector3
    var adaptingLuminance: Double = 20
    var backgroundLuminanceFactor: Double = 20
    var surroundF: Double = 1
    var surroundC: Double = 0.69
    var surroundNc: Double = 1
    var adaptationDegreeOverride: Double?
}

struct AppearanceCorrelates: Equatable, Sendable {
    let lightness: Double
    let brightness: Double
    let chroma: Double
    let colorfulness: Double
    let saturation: Double
    /// Undefined for achromatic colours, including black.
    let hue: Double?

    static let black = Self(
        lightness: 0, brightness: 0, chroma: 0,
        colorfulness: 0, saturation: 0, hue: nil
    )
}

/// CIE 248:2022, as specified in Gao et al., doi:10.1002/col.22959, Appendix A.
/// Includes the CIECAM16 linear extensions of the response function (not CAM16 2017).
struct CIECAM16: Sendable {
    let conditions: AppearanceViewingConditions
    let adaptationDegree: Double

    private let gains: Vector3
    private let luminanceFactor: Double
    private let luminanceRoot: Double
    private let nbb: Double
    private let z: Double
    private let whiteResponse: Double
    private let chromaFactor: Double
    private let upperResponseLimit: Double

    init(conditions: AppearanceViewingConditions) throws {
        let white = conditions.whiteXYZ
        guard [white.first, white.second, white.third,
               conditions.adaptingLuminance, conditions.backgroundLuminanceFactor,
               conditions.surroundF, conditions.surroundC, conditions.surroundNc]
            .allSatisfy(\.isFinite),
              white.first > 0, abs(white.second - 100) < 1e-9, white.third > 0,
              conditions.adaptingLuminance > 0,
              conditions.backgroundLuminanceFactor > 0,
              conditions.surroundF > 0, conditions.surroundC > 0,
              conditions.surroundNc > 0 else {
            throw ColorAppearanceError.invalidViewingConditions
        }
        let degree = conditions.adaptationDegreeOverride ?? min(1, max(0,
            conditions.surroundF * (1 - exp((-conditions.adaptingLuminance - 42) / 92) / 3.6)
        ))
        guard degree.isFinite, (0...1).contains(degree) else {
            throw ColorAppearanceError.invalidViewingConditions
        }
        let coneWhite = Self.coneResponse(white)
        guard coneWhite.first > 0, coneWhite.second > 0, coneWhite.third > 0 else {
            throw ColorAppearanceError.invalidViewingConditions
        }
        let gains = Self.map(coneWhite) { degree * 100 / $0 + 1 - degree }
        let adaptedWhite = Self.product(coneWhite, gains)
        let la = conditions.adaptingLuminance
        let k4 = pow(1 / (5 * la + 1), 4)
        let fl = 0.2 * k4 * 5 * la + 0.1 * pow(1 - k4, 2) * cbrt(5 * la)
        let n = conditions.backgroundLuminanceFactor / 100
        let nbb = 0.725 * pow(n, -0.2)
        // The adopted white uses the original compression, per Appendix A step 0.
        let compressedWhite = Self.map(adaptedWhite) { Self.response($0, fl: fl) + 0.1 }
        let aw = Self.achromaticResponse(compressedWhite, nbb: nbb)
        guard fl.isFinite, fl > 0, aw.isFinite, aw > 0 else {
            throw ColorAppearanceError.invalidViewingConditions
        }
        self.conditions = conditions
        adaptationDegree = degree
        self.gains = gains
        luminanceFactor = fl
        luminanceRoot = pow(fl, 0.25)
        self.nbb = nbb
        z = 1.48 + sqrt(n)
        whiteResponse = aw
        chromaFactor = pow(1.64 - pow(0.29, n), 0.73)
        upperResponseLimit = max(150, adaptedWhite.first, adaptedWhite.second, adaptedWhite.third)
    }

    func forward(_ xyz: Vector3) throws -> AppearanceCorrelates {
        guard Self.isFinite(xyz), xyz.second >= 0 else {
            throw ColorAppearanceError.invalidStimulus
        }
        if xyz.first == 0, xyz.second == 0, xyz.third == 0 { return .black }
        let adapted = Self.product(Self.coneResponse(xyz), gains)
        let compressed = Self.map(adapted) { compressedResponse($0) }
        let a = compressed.first - 12 * compressed.second / 11 + compressed.third / 11
        let b = (compressed.first + compressed.second - 2 * compressed.third) / 9
        let opponentMagnitude = hypot(a, b)
        let angle = atan2(b, a)
        let hue = (angle * 180 / .pi + 360).truncatingRemainder(dividingBy: 360)
        let achromatic = Self.achromaticResponse(compressed, nbb: nbb)
        let denominator = compressed.first + compressed.second + 21 * compressed.third / 20
        guard achromatic > 0, denominator > 0 else {
            throw ColorAppearanceError.outsideModelDomain
        }
        let j = 100 * pow(achromatic / whiteResponse, conditions.surroundC * z)
        let q = 4 / conditions.surroundC * sqrt(j / 100) * (whiteResponse + 4) * luminanceRoot
        let eccentricity = (cos(angle + 2) + 3.8) / 4
        let t = 50_000 / 13.0 * conditions.surroundNc * nbb
            * eccentricity * opponentMagnitude / denominator
        let c = pow(t, 0.9) * sqrt(j / 100) * chromaFactor
        let m = c * luminanceRoot
        let s = 100 * sqrt(m / q)
        guard [j, q, c, m, s, hue].allSatisfy(\.isFinite) else {
            throw ColorAppearanceError.outsideModelDomain
        }
        return AppearanceCorrelates(
            lightness: j, brightness: q, chroma: c,
            colorfulness: m, saturation: s, hue: c > 1e-10 ? hue : nil
        )
    }

    /// Reproduce Q, M and h together under these viewing conditions.
    func inverse(_ appearance: AppearanceCorrelates) throws -> Vector3 {
        try inverse(
            brightness: appearance.brightness,
            colorfulness: appearance.colorfulness,
            hue: appearance.hue
        )
    }

    func inverse(brightness q: Double, colorfulness m: Double, hue: Double?) throws -> Vector3 {
        guard q.isFinite, m.isFinite, q >= 0, m >= 0,
              hue?.isFinite != false else {
            throw ColorAppearanceError.invalidStimulus
        }
        if q == 0 {
            guard m == 0 else { throw ColorAppearanceError.outsideModelDomain }
            return Vector3(first: 0, second: 0, third: 0)
        }
        guard m <= 1e-10 || hue != nil else { throw ColorAppearanceError.invalidStimulus }
        let rootJ = q * conditions.surroundC / (4 * (whiteResponse + 4) * luminanceRoot)
        let c = m / luminanceRoot
        let t = m <= 1e-10 ? 0 : pow(c / (rootJ * chromaFactor), 1 / 0.9)
        let angle = (hue ?? 0) * .pi / 180
        let eccentricity = (cos(angle + 2) + 3.8) / 4
        let achromatic = whiteResponse * pow(rootJ * rootJ, 1 / (conditions.surroundC * z))
        let p2 = achromatic / nbb + 0.305
        let p1 = 50_000 / 13.0 * conditions.surroundNc * nbb * eccentricity
        // Algebraic solution of the opponent/achromatic equations; no hue-axis divisions.
        let denominator = 23 * p1 + t * (11 * cos(angle) + 108 * sin(angle))
        guard denominator > 0, denominator.isFinite else {
            throw ColorAppearanceError.outsideModelDomain
        }
        let radius = 23 * p2 * t / denominator
        let a = radius * cos(angle)
        let b = radius * sin(angle)
        let compressed = Vector3(
            first: (460 * p2 + 451 * a + 288 * b) / 1403,
            second: (460 * p2 - 891 * a - 261 * b) / 1403,
            third: (460 * p2 - 220 * a - 6300 * b) / 1403
        )
        let adapted = Self.map(compressed) { inverseCompressedResponse($0) }
        let cones = Vector3(
            first: adapted.first / gains.first,
            second: adapted.second / gains.second,
            third: adapted.third / gains.third
        )
        let xyz = Self.xyz(fromCones: cones)
        guard Self.isFinite(xyz) else { throw ColorAppearanceError.outsideModelDomain }
        return xyz
    }

    private func compressedResponse(_ value: Double) -> Double {
        let lower = 0.26
        if value <= lower {
            return Self.response(lower, fl: luminanceFactor) * value / lower + 0.1
        }
        if value >= upperResponseLimit {
            return Self.response(upperResponseLimit, fl: luminanceFactor)
                + responseDerivative(at: upperResponseLimit) * (value - upperResponseLimit) + 0.1
        }
        return Self.response(value, fl: luminanceFactor) + 0.1
    }

    private func inverseCompressedResponse(_ compressed: Double) -> Double {
        let value = compressed - 0.1
        let lower = Self.response(0.26, fl: luminanceFactor)
        let upper = Self.response(upperResponseLimit, fl: luminanceFactor)
        if value <= lower { return value * 0.26 / lower }
        if value >= upper {
            return upperResponseLimit + (value - upper) / responseDerivative(at: upperResponseLimit)
        }
        return 100 / luminanceFactor * pow(27.13 * value / (400 - value), 1 / 0.42)
    }

    private func responseDerivative(at value: Double) -> Double {
        let scaled = luminanceFactor * value / 100
        return 400 * 0.42 * 27.13 * luminanceFactor / 100 * pow(scaled, -0.58)
            / pow(27.13 + pow(scaled, 0.42), 2)
    }

    private static func response(_ value: Double, fl: Double) -> Double {
        let powered = pow(fl * abs(value) / 100, 0.42)
        return (value < 0 ? -1 : 1) * 400 * powered / (27.13 + powered)
    }

    private static func achromaticResponse(_ v: Vector3, nbb: Double) -> Double {
        (2 * v.first + v.second + v.third / 20 - 0.305) * nbb
    }

    static func coneResponse(_ v: Vector3) -> Vector3 {
        Vector3(
            first: 0.401288 * v.first + 0.650173 * v.second - 0.051461 * v.third,
            second: -0.250268 * v.first + 1.204414 * v.second + 0.045854 * v.third,
            third: -0.002079 * v.first + 0.048952 * v.second + 0.953127 * v.third
        )
    }

    private static func xyz(fromCones v: Vector3) -> Vector3 {
        Vector3(
            first: 1.8620678550872327 * v.first - 1.0112546305316843 * v.second + 0.14918677544445175 * v.third,
            second: 0.3875265432361372 * v.first + 0.6214474419314753 * v.second - 0.008973985167612518 * v.third,
            third: -0.01584149884933386 * v.first - 0.03412293802851557 * v.second + 1.0499644368778496 * v.third
        )
    }

    private static func product(_ a: Vector3, _ b: Vector3) -> Vector3 {
        Vector3(first: a.first * b.first, second: a.second * b.second, third: a.third * b.third)
    }

    private static func map(_ v: Vector3, _ operation: (Double) -> Double) -> Vector3 {
        Vector3(first: operation(v.first), second: operation(v.second), third: operation(v.third))
    }

    private static func isFinite(_ v: Vector3) -> Bool {
        v.first.isFinite && v.second.isFinite && v.third.isFinite
    }
}

File: Windows/src/IwashiScope.Core/Calculations/CieCam16.cs
using IwashiScope.Core.Models;

namespace IwashiScope.Core.Calculations;

public enum ColorAppearanceError
{
    InvalidViewingConditions,
    InvalidStimulus,
    OutsideModelDomain,
    InsufficientSpectrum,
}

public sealed class ColorAppearanceException(ColorAppearanceError error) : Exception(error switch
{
    ColorAppearanceError.InvalidViewingConditions => "The color appearance viewing conditions are invalid.",
    ColorAppearanceError.InvalidStimulus => "The colorimetric input for color appearance is invalid.",
    ColorAppearanceError.OutsideModelDomain => "This color is outside the appearance model's calculation domain.",
    _ => "There is insufficient spectral coverage for this comparison.",
})
{
    public ColorAppearanceError Error { get; } = error;
}

/// <summary>White XYZ and Yb use Yw = 100. LA is absolute luminance, not XYZ Y.</summary>
public sealed record AppearanceViewingConditions(
    Vector3 WhiteXyz,
    double AdaptingLuminance = 20,
    double BackgroundLuminanceFactor = 20,
    double SurroundF = 1,
    double SurroundC = 0.69,
    double SurroundNc = 1,
    double? AdaptationDegreeOverride = null);

public sealed record AppearanceCorrelates(
    double Lightness,
    double Brightness,
    double Chroma,
    double Colorfulness,
    double Saturation,
    double? Hue)
{
    public static readonly AppearanceCorrelates Black = new(0, 0, 0, 0, 0, null);
}

/// <summary>
/// CIE 248:2022, Gao et al. doi:10.1002/col.22959 Appendix A.
/// Uses the CIECAM16 response extensions, not the older CAM16 (2017) response.
/// Matches the macOS implementation and its independent Colour 0.4.7 fixtures.
/// </summary>
public sealed class CieCam16
{
    public AppearanceViewingConditions Conditions { get; }
    public double AdaptationDegree { get; }
    private readonly Vector3 _gains;
    private readonly double _luminanceFactor;
    private readonly double _luminanceRoot;
    private readonly double _nbb;
    private readonly double _z;
    private readonly double _whiteResponse;
    private readonly double _chromaFactor;
    private readonly double _upperResponseLimit;

    public CieCam16(AppearanceViewingConditions conditions)
    {
        var white = conditions.WhiteXyz;
        if (white is not { IsFinite: true } || white.First <= 0 || white.Third <= 0 ||
            Math.Abs(white.Second - 100) >= 1e-9 ||
            !Finite(conditions.AdaptingLuminance, conditions.BackgroundLuminanceFactor,
                conditions.SurroundF, conditions.SurroundC, conditions.SurroundNc) ||
            conditions.AdaptingLuminance <= 0 || conditions.BackgroundLuminanceFactor <= 0 ||
            conditions.SurroundF <= 0 || conditions.SurroundC <= 0 || conditions.SurroundNc <= 0)
        {
            throw new ColorAppearanceException(ColorAppearanceError.InvalidViewingConditions);
        }
        var degree = conditions.AdaptationDegreeOverride ?? Math.Clamp(
            conditions.SurroundF * (1 - Math.Exp((-conditions.AdaptingLuminance - 42) / 92) / 3.6), 0, 1);
        var coneWhite = ConeResponse(white);
        if (!double.IsFinite(degree) || degree is < 0 or > 1 ||
            coneWhite.First <= 0 || coneWhite.Second <= 0 || coneWhite.Third <= 0)
        {
            throw new ColorAppearanceException(ColorAppearanceError.InvalidViewingConditions);
        }
        var gains = Map(coneWhite, value => degree * 100 / value + 1 - degree);
        var adaptedWhite = Product(coneWhite, gains);
        var la = conditions.AdaptingLuminance;
        var k4 = Math.Pow(1 / (5 * la + 1), 4);
        var fl = 0.2 * k4 * 5 * la + 0.1 * Math.Pow(1 - k4, 2) * Math.Cbrt(5 * la);
        var n = conditions.BackgroundLuminanceFactor / 100;
        var nbb = 0.725 * Math.Pow(n, -0.2);
        // The adopted white uses the original compression (Appendix A, step 0).
        var compressedWhite = Map(adaptedWhite, value => Response(value, fl) + 0.1);
        var aw = AchromaticResponse(compressedWhite, nbb);
        if (!Finite(fl, aw) || fl <= 0 || aw <= 0)
        {
            throw new ColorAppearanceException(ColorAppearanceError.InvalidViewingConditions);
        }
        Conditions = conditions;
        AdaptationDegree = degree;
        _gains = gains;
        _luminanceFactor = fl;
        _luminanceRoot = Math.Pow(fl, 0.25);
        _nbb = nbb;
        _z = 1.48 + Math.Sqrt(n);
        _whiteResponse = aw;
        _chromaFactor = Math.Pow(1.64 - Math.Pow(0.29, n), 0.73);
        _upperResponseLimit = Math.Max(150,
            Math.Max(adaptedWhite.First, Math.Max(adaptedWhite.Second, adaptedWhite.Third)));
    }

    public AppearanceCorrelates Forward(Vector3 xyz)
    {
        if (xyz is not { IsFinite: true } || xyz.Second < 0)
        {
            throw new ColorAppearanceException(ColorAppearanceError.InvalidStimulus);
        }
        if (xyz == new Vector3(0, 0, 0)) return AppearanceCorrelates.Black;
        var adapted = Product(ConeResponse(xyz), _gains);
        var compressed = Map(adapted, CompressedResponse);
        var a = compressed.First - 12 * compressed.Second / 11 + compressed.Third / 11;
        var b = (compressed.First + compressed.Second - 2 * compressed.Third) / 9;
        var angle = Math.Atan2(b, a);
        var hue = (angle * 180 / Math.PI + 360) % 360;
        var achromatic = AchromaticResponse(compressed, _nbb);
        var denominator = compressed.First + compressed.Second + 21 * compressed.Third / 20;
        if (achromatic <= 0 || denominator <= 0)
        {
            throw new ColorAppearanceException(ColorAppearanceError.OutsideModelDomain);
        }
        var j = 100 * Math.Pow(achromatic / _whiteResponse, Conditions.SurroundC * _z);
        var q = 4 / Conditions.SurroundC * Math.Sqrt(j / 100) * (_whiteResponse + 4) * _luminanceRoot;
        var eccentricity = (Math.Cos(angle + 2) + 3.8) / 4;
        var t = 50_000 / 13.0 * Conditions.SurroundNc * _nbb
            * eccentricity * Hypot(a, b) / denominator;
        var c = Math.Pow(t, 0.9) * Math.Sqrt(j / 100) * _chromaFactor;
        var m = c * _luminanceRoot;
        var s = 100 * Math.Sqrt(m / q);
        if (!Finite(j, q, c, m, s, hue))
        {
            throw new ColorAppearanceException(ColorAppearanceError.OutsideModelDomain);
        }
        return new AppearanceCorrelates(j, q, c, m, s, c > 1e-10 ? hue : null);
    }

    public Vector3 Inverse(AppearanceCorrelates appearance) =>
        Inverse(appearance.Brightness, appearance.Colorfulness, appearance.Hue);

    /// <summary>Reproduces the source's Q, M and h as one set in these conditions.</summary>
    public Vector3 Inverse(double brightness, double colorfulness, double? hue)
    {
        var q = brightness;
        var m = colorfulness;
        if (!Finite(q, m) || q < 0 || m < 0 || hue is { } h && !double.IsFinite(h))
        {
            throw new ColorAppearanceException(ColorAppearanceError.InvalidStimulus);
        }
        if (q == 0)
        {
            if (m != 0) throw new ColorAppearanceException(ColorAppearanceError.OutsideModelDomain);
            return new Vector3(0, 0, 0);
        }
        if (m > 1e-10 && hue is null)
        {
            throw new ColorAppearanceException(ColorAppearanceError.InvalidStimulus);
        }
        var rootJ = q * Conditions.SurroundC / (4 * (_whiteResponse + 4) * _luminanceRoot);
        var c = m / _luminanceRoot;
        var t = m <= 1e-10 ? 0 : Math.Pow(c / (rootJ * _chromaFactor), 1 / 0.9);
        var angle = (hue ?? 0) * Math.PI / 180;
        var eccentricity = (Math.Cos(angle + 2) + 3.8) / 4;
        var achromatic = _whiteResponse * Math.Pow(rootJ * rootJ, 1 / (Conditions.SurroundC * _z));
        var p2 = achromatic / _nbb + 0.305;
        var p1 = 50_000 / 13.0 * Conditions.SurroundNc * _nbb * eccentricity;
        var denominator = 23 * p1 + t * (11 * Math.Cos(angle) + 108 * Math.Sin(angle));
        if (!double.IsFinite(denominator) || denominator <= 0)
        {
            throw new ColorAppearanceException(ColorAppearanceError.OutsideModelDomain);
        }
        var radius = 23 * p2 * t / denominator;
        var a = radius * Math.Cos(angle);
        var b = radius * Math.Sin(angle);
        var compressed = new Vector3(
            (460 * p2 + 451 * a + 288 * b) / 1403,
            (460 * p2 - 891 * a - 261 * b) / 1403,
            (460 * p2 - 220 * a - 6300 * b) / 1403);
        var adapted = Map(compressed, InverseCompressedResponse);
        var cones = new Vector3(adapted.First / _gains.First,
            adapted.Second / _gains.Second, adapted.Third / _gains.Third);
        var xyz = XyzFromCones(cones);
        if (!xyz.IsFinite) throw new ColorAppearanceException(ColorAppearanceError.OutsideModelDomain);
        return xyz;
    }

    private double CompressedResponse(double value)
    {
        const double lower = 0.26;
        if (value <= lower) return Response(lower, _luminanceFactor) * value / lower + 0.1;
        if (value >= _upperResponseLimit)
            return Response(_upperResponseLimit, _luminanceFactor)
                + ResponseDerivative(_upperResponseLimit) * (value - _upperResponseLimit) + 0.1;
        return Response(value, _luminanceFactor) + 0.1;
    }

    private double InverseCompressedResponse(double compressed)
    {
        var value = compressed - 0.1;
        var lower = Response(0.26, _luminanceFactor);
        var upper = Response(_upperResponseLimit, _luminanceFactor);
        if (value <= lower) return value * 0.26 / lower;
        if (value >= upper) return _upperResponseLimit + (value - upper) / ResponseDerivative(_upperResponseLimit);
        return 100 / _luminanceFactor * Math.Pow(27.13 * value / (400 - value), 1 / 0.42);
    }

    private double ResponseDerivative(double value)
    {
        var scaled = _luminanceFactor * value / 100;
        return 400 * 0.42 * 27.13 * _luminanceFactor / 100 * Math.Pow(scaled, -0.58)
            / Math.Pow(27.13 + Math.Pow(scaled, 0.42), 2);
    }

    private static double Response(double value, double fl)
    {
        var powered = Math.Pow(fl * Math.Abs(value) / 100, 0.42);
        return (value < 0 ? -1 : 1) * 400 * powered / (27.13 + powered);
    }

    private static double AchromaticResponse(Vector3 v, double nbb) =>
        (2 * v.First + v.Second + v.Third / 20 - 0.305) * nbb;

    public static Vector3 ConeResponse(Vector3 v) => new(
        0.401288 * v.First + 0.650173 * v.Second - 0.051461 * v.Third,
        -0.250268 * v.First + 1.204414 * v.Second + 0.045854 * v.Third,
        -0.002079 * v.First + 0.048952 * v.Second + 0.953127 * v.Third);

    private static Vector3 XyzFromCones(Vector3 v) => new(
        1.8620678550872327 * v.First - 1.0112546305316843 * v.Second + 0.14918677544445175 * v.Third,
        0.3875265432361372 * v.First + 0.6214474419314753 * v.Second - 0.008973985167612518 * v.Third,
        -0.01584149884933386 * v.First - 0.03412293802851557 * v.Second + 1.0499644368778496 * v.Third);

    private static Vector3 Product(Vector3 a, Vector3 b) =>
        new(a.First * b.First, a.Second * b.Second, a.Third * b.Third);
    private static Vector3 Map(Vector3 v, Func<double, double> operation) =>
        new(operation(v.First), operation(v.Second), operation(v.Third));
    private static bool Finite(params double[] values) => values.All(double.IsFinite);
    private static double Hypot(double x, double y)
    {
        var maximum = Math.Max(Math.Abs(x), Math.Abs(y));
        return maximum == 0 ? 0 : maximum * Math.Sqrt(1 + Math.Pow(Math.Min(Math.Abs(x), Math.Abs(y)) / maximum, 2));
    }
}
